### ISSUE TITLE
Warp/Mesh Transformation

### ISSUE THREAD

**Issue by panhartstuff** [drawpile/Drawpile#609]
A feature I've been dying to see is the ability to "Distort" and "Warp" selection like in Photoshop. It's super useful for fixing proportions and slanted drawings

Here's an example of what I mean
Distort (called "Free Transform" in Clip Studio Paint):
![2018-05-01_22-19-45](https://user-images.githubusercontent.com/34397257/39478943-fc3e9754-4d8d-11e8-92c9-4a8c1699a46b.gif)

Warp (called "Mesh Transformation" in Clip Studio Paint):
![2018-05-01_22-21-50](https://user-images.githubusercontent.com/34397257/39478973-17fac9ea-4d8e-11e8-827c-e23619f58f3b.gif)

Please do consider developing these features!

**Comment by Waloumi** [comment]
I very much agree.

**Comment by callaa** [comment]
It's not quite as versatile as the free transform mode, but a selection shear feature already exists. (Hold  Shift+Alt and drag)

**Comment by panhartstuff** [comment]
Any updates on this?

**Comment by callaa** [comment]
No, haven't gotten around to this yet. (If anyone else wishes to tackle it in the meantime, I welcome the help.)

**Comment by askmeaboutlo0m** [comment]
I've implemented the Distort aka Free Transform aka Free Deform feature in [this branch](https://github.com/askmeaboutlo0m/Drawpile/tree/transformplus). I also made [a small recording to show it off](https://cdn.discordapp.com/attachments/543731850691215361/742020351520669707/distort.webm).

I kinda still want to fix that the handles aren't in the right positions though. And maybe prevent constellations that cause very strange effects... although I dunno if that's worth it, since if you're making silly transformations, that's the result you get.

Warp aka Mesh Transform seems to be a whole lot more complicated though and not something I've ever used myself, so I'm not sure if I'll get around to implementing that as well.

**Comment by MorrowShore** [comment]
Partially implemented as transform tools.

**Comment by cromachina** [comment]
Could also add a simple "linear distort" option for the main transform tool along side this. I think that this is the same algorithm as other applications' "warp" tools (2D affine transformation?). Linear distortion is usually more desirable than perspective distortion when tweaking drawings, and perspective distortion makes more sense when actually trying to warp something specifically into perspective.

Here is a video of that in SAI, which doesn't have mesh transform, but does have linear distortion on a sliding scale (which shows that there is some interpolated transformation between linear distortion and perspective distortion):

https://github.com/user-attachments/assets/ca716897-9379-4ea6-b2a9-570f3faf5185

**Comment by askmeaboutlo0m** [comment]
Adding such an option probably wouldn't be difficult conceptually, but I don't know how the math for this works at all. What we use is based on Qt's `QTransform::quadToQuad` function, which in turn transforms the source quad to a unit square and then the unit square to the target quad. I can find plenty of sources on algorithms for that, but none that document how the math would work for a linear transformation.

**Comment by cromachina** [comment]
I'm not sure if the linear transform can actually be done with a matrix transform, but I think it works like a series of linear interpolations.
Given an orthogonal aligned rectangle `pqst` (the selection), and a target set of control points that form a quadrilateral `p'q's't'`, a point in the input rectangle `w` is mapped to the quadrilateral as `w'` by finding the proportional factor of each coordinate of `w` in `pqst`. You then construct vectors from similar positions in `p'q's't'`. `a'` and `c'` share the same factor as `a` and `c`, and then the factor for `d` finds where on `a'c'` that `w'` lies.

![Image](https://github.com/user-attachments/assets/0f1e1f4f-27ec-4d12-a33e-a1225e9d555c)

**Comment by cromachina** [comment]
Actually it's probably just bilinear interpolation, but I'm not sure if it would work with Qt's transform functions for 2D images. https://en.wikipedia.org/wiki/Bilinear_interpolation#Alternative_matrix_form

**Comment by askmeaboutlo0m** [comment]
Not being able to go through a matrix transform is pretty obstructive, since it means that it can't be previewed through Qt (the "fast" preview mode), or at least not without doing major patching on it. It's still possible to do it in the accurate mode, since we have a custom implementation for that. But I'd need actual code to base it off of anyway, going off of the raw math with no oracle to compare it to would take excessive amounts of time, probably longer than it would take to port Krita's warp transform. Which may happen to include this kind of linear transform algorithm as a side-effect by setting the subdivision to 2 and the anchor strength to affine. At least it looks very similar to the SAI result with 0 perspective to me.
